ethereum/sidecar/attestation_validation.go

```diff
@@ -0,0 +1,156 @@
+package sidecar
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	ethconfig "github.com/keep-network/keep-common/pkg/chain/ethereum"
+	ethconnect "github.com/mezo-org/mezod/ethereum"
+	"github.com/mezo-org/mezod/ethereum/bindings/portal"
+)
+
+var (
+	ErrInvalidAttestation      = errors.New("invalid attestation")
+	ErrValidatorNotInTheBitmap = errors.New("validatorId not in the bitmap yet")
+)
+
+type AttestationValidation struct {
+	bridgeContract ethconnect.BridgeContract
+	address        common.Address
+}
+
+func NewAttestationValidation(
+	bridgeContract ethconnect.BridgeContract,
+	validatorAddress common.Address,
+) *AttestationValidation {
+	return &AttestationValidation{
+		bridgeContract: bridgeContract,
+		address:        validatorAddress,
+	}
+}
+
+func (av *AttestationValidation) IsConfirmed(
+	attestation *portal.MezoBridgeAssetsUnlocked,
+) error {
+	ok, err := av.bridgeContract.ValidateAssetsUnlocked(*attestation)
+	if err != nil {
+		return fmt.Errorf("couldn't call validateAssetsUnlocked: %w", err)
+	}
+	if !ok {
+		// this is a specific case of error
+		// the attestation is not valid, we specify it down the line
+		return ErrInvalidAttestation
+	}
+
+	ok, err = av.bridgeContract.ConfirmedUnlocks(attestation.Amount)
+	if err != nil {
+		return fmt.Errorf("couldn't get confirmedLocks: %w", err)
+	}
+	if ok {
+		return nil
+	}
+
+	return av.checkOwnConfirmation(attestation)
+}
+
+func (av *AttestationValidation) WaitForConfirmation(
+	blockHeightWaiter ethconfig.BlockHeightWaiter,
+	startBlock, confirmations uint64,
+	attestation *portal.MezoBridgeAssetsUnlocked,
+) error {
+	return WaitForBlockConfirmations(
+		blockHeightWaiter,
+		startBlock,
+		confirmations,
+		func() error {
+			return av.checkOwnConfirmation(attestation)
+		},
+	)
+}
+
+func (av *AttestationValidation) checkOwnConfirmation(
+	attestation *portal.MezoBridgeAssetsUnlocked,
+) error {
+	encoded, err := abiEncodeAttestation(attestation)
+	if err != nil {
+		return fmt.Errorf("couldn't ABI encode attestation: %w", err)
+	}
+
+	hash := crypto.Keccak256Hash(encoded)
+
+	bitmap, err := av.bridgeContract.Attestations(hash)
+	if err != nil {
+		return fmt.Errorf("couldn't get confirmedLock: %w", err)
+	}
+
+	validatorID, err := av.bridgeContract.ValidatorIDs(av.address)
+	if err != nil {
+		return fmt.Errorf("couldn't get validator ID: %w", err)
+	}
+
+	mask := new(big.Int).Lsh(big.NewInt(1), uint(validatorID))
+
+	if new(big.Int).And(bitmap, mask).Int64() != 0 {
```

Comment:
If this condition is true, the validator already attested, see: https://github.com/thesis/mezo-portal/blob/aa3d044e7ff26110833fdb4c7a7a7bb4baeb0e50/solidity/contracts/MezoBridge.sol#L305. I believe we should do `== 0` here.

Reply:
a6b3cb22

Reply:
Wrong commit hash?

Reply:
Nope it should be fine? a6b3cb227524057d08d360350bb01cafd839c7d2